### ISSUE TITLE
Make ceph-csi domainLabels usable with Rook

### ISSUE THREAD

**Issue by sbskas** [ceph/ceph-csi#2962]
# Describe the feature you'd like to have #
At the present time, to use topologyContraintedPools, the rbdplugin needs to be given a list of attributes
fo sort the different pools to be used.

Problem is this attributes list must be given on the cmdline.
Rook is unable to customize the attribute cmdline of the ceph-rbdplugin.
To pass this attribute, one has disable the rbd Daemonset managed by rook and create a new helm deployment
of ceph-csi defeating the seamless integration with rook.

My proposal is to pass the value in domainLabels not on the cmdline but as an atrribute in the ceph-csi-config
It may look like this:
```
csi-cluster-config-json: [
  {
    "clusterID":"rook-ceph",
    "monitors":[
        "10.1.0.10:6789",
        "10.2.0.10:6789",
        "10.3.0.10:6789"
    ],
    domainLabels: [
       "topology.kubernetes.io/zone",
       "kubernetes.io/hostname"
    ]
  }
]'
```
# What is the value to the end user? (why is it a priority?) #

This parameter needs to be integrated in config files rather than on a cmdline swtitch.
Also, one could have different domainLabels list for different cluster.

# How will we know we have a good solution? (acceptance criteria) #

We'll know if we're able to pass the domainLabels without touching the daemonSet definition.

**Comment by Madhu-1** [comment]
cc @ShyamsundarR

**Comment by Madhu-1** [comment]
>Rook is unable to customize the attribute cmdline of the ceph-rbdplugin.

Today this is not possible but this is possible to code changes in Rook to parse the value from Rook configmap and update the daemonset pod.

>To pass this attribute, one has disable the rbd Daemonset managed by rook and create a new helm deployment
of ceph-csi defeating the seamless integration with rook.

Why disable is required. Rook handles the change in configuration or new configuration seamlessly today as well for other configurations also.

> This parameter needs to be integrated in config files rather than on a cmdline swtitch.
Also, one could have different domainLabels list for different cluster.

@ShyamsundarR any thoughts on this one. this is going to be a design change of original implementation.

**Comment by humblec** [comment]
@sbskas the entries from ceph cluster configmap and making Rook aware of the configuration and acting on  is one thought . 

One other thought from my end is that, keeping a seperate configmap ( not as part of cluster configmap)  for topology alone ( may be keeping as a map of k:v of nodes and topology info ) then making use of it in any layer independently . Say Rook can fetch and use it, or even in absence of Rook, other layer should be able to consume and use ...etc.

**Comment by Rakshith-R** [comment]
I think we have a specific set of topology labels https://rook.io/docs/rook/latest/ceph-cluster-crd.html#osd-topology 
Maybe we can by default pass the labels keys to cephcsi through cmdline args instead of changing the way we pass it.

Anyway, this will be used only when storageclass has topology options https://github.com/ceph/ceph-csi/blob/f6ae6120036a9dbb8903f70f44f4194138abe906/examples/rbd/storageclass.yaml#L119

**Comment by sbskas** [comment]
https://github.com/rook/rook/blob/656e9466d4cee426c9772026822ac881421dd542/pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml#L59-L73
@Madhu-1  Nothing in the rook csi-rbdplugin template allows to pass the expected parameter.
If you make rook ceph-csi "topology aware", you introduce more coupling between the two projects.
If you change the topology implementation in ceph-csi rbdplugin, you have to take care of not breaking the rook support code of it. The reverse is also true.
If you put it in an opaque config-map (let's say in the ceph-csi-config that holds the cluster monitors definition), you avoid this coupling. It's just a parameter to customize in a configMap.
@humblec keeping 2 different configmap is also ok, it's just a proposal to discuss.

**Comment by Rakshith-R** [comment]
> https://github.com/rook/rook/blob/656e9466d4cee426c9772026822ac881421dd542/pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml#L59-L73 @Madhu-1 Nothing in the rook csi-rbdplugin template allows to pass the expected parameter. If you make rook ceph-csi "topology aware", you introduce more coupling between the two projects. If you change the topology implementation in ceph-csi rbdplugin, you have to take care of not breaking the rook support code of it. The reverse is also true. If you put it in an opaque config-map (let's say in the ceph-csi-config that holds the cluster monitors definition), you avoid this coupling. It's just a parameter to customize in a configMap. 

Rook will reconcile the configMap and the domainlabel values will be wiped. 
https://github.com/rook/rook/blob/656e9466d4cee426c9772026822ac881421dd542/pkg/operator/ceph/csi/cluster_config.go#L81

**Comment by sbskas** [comment]
@Rakshith-R 
My concern for now is the ceph-csi code.
Rook code would be amended in a different issue (a rook one).
Looking a the code, the option could be handled like the RadosNamespace option.

**Comment by Rakshith-R** [comment]
> @Rakshith-R 
> My concern for now is the ceph-csi code.
> Rook code would be amended in a different issue (a rook one).
> Looking a the code, the option could be handled like the RadosNamespace option.


Do you expect to use different domain Labels keys with different cluster ids?

I assume not, since the label keys will remain the same, I recommend we code them into Rook rbdplugin daemonset cmdline.
Again this will come into effect only when topology constraints is used in storageclass.

This will need minimal code changes and effort. 

(+ I prefer not to over complicate the way rook handle ceph-csi configmap)

**Comment by sbskas** [comment]
We do not expect to have more than cluster running on our side, but, that's the initial plan.
However, rook is supporting several cluster, ceph-csi is also supporting multiple clusters.
domainLabels support only one list of labels which should be compatible with all deployed clusters.

**Comment by sbskas** [comment]
There is 3 paths from here:
1. Do nothing and take care of this thing in the rook code with the shortcomings I mentionned previously.
   To date, the code for handling such case in rook remains to be written.
2. Try to figure out the domainLabel list from the topologyconstrainedpools given in the storageclass.
    This would have the great advantage to be self-sufficient (no extra code needed) and relies on the ceph-csi code only.
3. Use this proposal or something closely related per cluster.

I don't like the proposal 1 for it leads to put configuration on the cmdline, meaning it needs another configuration part in the rook configuration.
Proposal 2 could be interesting. ceph is already following a defacto list of priorites in the crushmap : zone -> datacenter -> ... If one can map the node labels to this list, the domainLabels could be infered from the storageclass.
Also, those domainLabels order could be deduced from the crushmap rule associated to the pool, thus avoiding the redondant definition of domainLabels.

Proposal 3 was already discuted earlier, and I won't detail it further.

**Comment by ShyamsundarR** [comment]
Revisiting history, I am not sure why the `domainLabels` were added as a CLI option and not part of the configMap [1]:
  - I tried reasoning this again based on this being CO specific (i.e k8s specific) and maybe that was the reason to not add it!
  - Further attempted reasoning if we would have the said ConfigMap in time to populate CSIDriver with the required topology information, it does seem so [2]
  - There was a limitation that once a node is added with a topology constraint label, the CSI nodeplugin needs to be restarted to pick it up and advertise its CSIDriver information again (unless we can now dynamically update CSIDriver information without restarting), this possibly led to the CLI than the ConfigMap, as that forces the restart

As a result, I would assume moving this to the config map is a reasonable option. I do not see the above problems as limiting the approach. I am not evaluating if this makes it easier to integrate with Rook as such, so the benefits of doing the same has to be evaluated.

I would like to point out 2 (related) gotcha's in the discussions above:
  - k8s node topology and storage topology are different entities, so we should not leverage Ceph CRUSH topology as such into k8s, there has to exist a mapping and that is governed by node labels, so we rely on the same
    - For example proposal (2) above possibly assumes a node is labelled with the same topology segment values as the CRUSH map and that may not be true. As k8s itself relies on node labels for pod scheduling and topology constrained provisioning and staging, we should remain in the same space with CSI
  - I believe we should not be limited to a Ceph cluster and that the topology (as it really refers to k8s nodes topology) should be common across all CSI npde plugins
    - Which hence implies a new section in the configMap (or a separate configMap that is populated by an init container, thereby freeing CSI code to be dependent on k8s as such)

[1] [Determining node domain by the CSI nodeplugin](https://github.com/ShyamsundarR/ceph-csi/blob/31b0766eb62b27308e69f4f0ffe8c37dca9c9a70/docs/design/proposals/TopologySupport.md#determining-node-domain-by-the-csi-nodeplugin)
[2] [CSIDriver population](https://github.com/ceph/ceph-csi/blob/c245436ec4f89e6801a21782da64aaa0d370070b/internal/rbd/driver/driver.go#L137)

**Comment by sbskas** [comment]
Ok, new section in configmap that will be then.
Should something like that be suitable ?
```
csi-cluster-config-json: [
  {
    "clusterID":"rook-ceph",
    "monitors":[
        "10.1.0.10:6789",
        "10.2.0.10:6789",
        "10.3.0.10:6789"
    ]
  }
]'
csi-topology-config.json: [
  {
    "clusterID": "rook-ceph"
     domainLabels: [
       "topology.kubernetes.io/zone",
       "kubernetes.io/hostname"
    ]
  }
]
```

**Comment by ShyamsundarR** [comment]
> Ok, new section in configmap that will be then. Should something like that be suitable ?
> 
> ```
> csi-cluster-config-json: [
> csi-topology-config.json: [
>   {
>     "clusterID": "rook-ceph"
> ```

Why add the `clusterID` here, these are node labels that would be read and reflected as Ceph-CSI node labels, so they are common across any Ceph cluster. Unless I am missing something, the `clusterID` is not required in this section.

**Comment by sbskas** [comment]
clusterId prefix allow to have different domainLabels order with different clusters.

If all clusters could rely on the same domainLabels order, there is no reason to use different priorities for the domainLabels.

However, I don't know if this is true for every cluster managed by the ceph-csi.

I can't think of a use case, which does not mean it doesn't exists.

Anyway, with or without clusterId is ok with me.
